### PR TITLE
[RA2 Ch3] Device Plugin operation and example

### PR DESCRIPTION
**Pull request by petorre**
From @arkadykanevsky and me towards closing #2255 .

### CODE REVIEW

**Review by arkadykanevsky — APPROVED** [review]

File: doc/ref_arch/kubernetes/chapters/chapter03.md [hunk 1]
```diff
@@ -211,6 +211,17 @@ Memory or Huge Pages are not considered by the Topology Manager. This can be don
 
 [Device Plugin Framework](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/) advertises device hardware resources to kubelet with which vendors can implement plugins for devices that may require vendor-specific activation and life cycle management, and securely maps these devices to containers.
 
+Figure 3-2 shows in four steps how device plugins operate on a Kubernetes node:
+* 1: During setup, the cluster administrator as human or operator in software (more in [3.2.1.8 Operator Pattern](chapter03.md#3218-operator-pattern)) knows or discovers (as per [3.2.1.5 Node Feature Discovery](chapter03.md#3215-node-feature-discovery)) what kind of devices are present on the different machines, selects which devices to enable and deploys the associated device plugins.
```
Comment (pgoyal01) [inline doc/ref_arch/kubernetes/chapters/chapter03.md:215]: @petorre @arkadykanevsky 
I don't think "human or software" is necessary
In other bullets, "node" is used
```suggestion
* 1: During setup, the cluster administrator (more in [3.2.1.8 Operator Pattern](chapter03.md#3218-operator-pattern)) knows or discovers (as per [3.2.1.5 Node Feature Discovery](chapter03.md#3215-node-feature-discovery)) what kind of devices are present on the different nodes, selects which devices to enable and deploys the associated device plugins.
```

File: doc/ref_arch/kubernetes/chapters/chapter03.md [hunk 2]
```diff
@@ -211,6 +211,17 @@ Memory or Huge Pages are not considered by the Topology Manager. This can be don
...
+* 3: A user submits a pod specification (workload manifest file) requesting a certain type of device.
+* 4: Kubelet decides which node to schedule workload on and which device on the node to assign to the pod's containers.
+
+<p align="center"><img src="../figures/Ch3_Figure_Device_Plugin_operation.png" alt="Device Plugin Operation" Title="Device Plugin Operation" width="50%"/></p>
+<p align="center"><b>Figure 3-2:</b> Device Plugin Operation</p>
+
+An example of often used device plugin is [SR-IOV Network Device Plugin](https://github.com/k8snetworkplumbingwg/sriov-network-device-plugin) that discovers and advertises SR-IOV Virtual Functions (VFs) available on a Kubernetes node, so is used to map VFs to scheduled pods. To use it needed are CNI meta plugin supporting network provisioning based on device plugin (like [Multus CNI](https://github.com/k8snetworkplumbingwg/multus-cni), to have additional network interface for VFs beyond original network interface for Kubernetes control) and SR-IOV CNI (During pod creation, plumbs allocated SR-IOV VF to a pods network namespace using VF information given by the meta plugin. On pod deletion, CNI plugin releases the VF from the pod.)
```
Comment (pgoyal01) [inline doc/ref_arch/kubernetes/chapters/chapter03.md:223]: typo
```suggestion
An example of often used device plugin is [SR-IOV Network Device Plugin](https://github.com/k8snetworkplumbingwg/sriov-network-device-plugin) that discovers and advertises SR-IOV Virtual Functions (VFs) available on a Kubernetes node, and is used to map VFs to scheduled pods. To use it, a CNI meta plugin, supporting network provisioning based on device plugin (like [Multus CNI](https://github.com/k8snetworkplumbingwg/multus-cni), to enable additional network interfaces for VFs (beyond the original network interface for Kubernetes control) and SR-IOV CNI. During pod creation, plumbs allocated SR-IOV VF to a pods network namespace using VF information given by the meta plugin. On pod deletion, CNI plugin releases the VF from the pod.
```

File: doc/ref_arch/kubernetes/chapters/chapter03.md [hunk 3]
```diff
@@ -220,7 +220,7 @@ Figure 3-2 shows in four steps how device plugins operate on a Kubernetes node:
...
+An example of often used device plugin is [SR-IOV Network Device Plugin](https://github.com/k8snetworkplumbingwg/sriov-network-device-plugin) that discovers and advertises SR-IOV Virtual Functions (VFs) available on a Kubernetes node, and is used to map VFs to scheduled pods. To use it, a CNI meta plugin, supporting network provisioning based on device plugin (like [Multus CNI](https://github.com/k8snetworkplumbingwg/multus-cni), to enable additional network interfaces for VFs (beyond the original network interface for Kubernetes control) and SR-IOV CNI. During pod creation, plumbs allocated SR-IOV VF to a pods network namespace using VF information given by the meta plugin. On pod deletion, CNI plugin releases the VF from the pod.
```
Comment (CsatariGergely) [inline doc/ref_arch/kubernetes/chapters/chapter03.md:223]:
```suggestion
An example of often used device plugin is [SR-IOV Network Device Plugin](https://github.com/k8snetworkplumbingwg/sriov-network-device-plugin) that discovers and advertises SR-IOV Virtual Functions (VFs) available on a Kubernetes node, and is used to map VFs to scheduled pods. To use it, a CNI meta plugin, supporting network provisioning based on device plugin (like [Multus CNI](https://github.com/k8snetworkplumbingwg/multus-cni) or [DANM](https://github.com/nokia/danm)) and SR-IOV CNI, to enable additional network interfaces for VFs (beyond the original network interface for Kubernetes control), are required. During pod creation, plumbs allocated SR-IOV VF to a pods network namespace using VF information given by the meta plugin. On pod deletion, CNI plugin releases the VF from the pod.
```

**Review by CsatariGergely — CHANGES_REQUESTED** [review]

**Review by CsatariGergely — APPROVED** [review]

File: doc/ref_arch/kubernetes/chapters/chapter03.md [hunk 4]
```diff
@@ -212,15 +212,15 @@ Memory or Huge Pages are not considered by the Topology Manager. This can be don
...
+An example of often used device plugin is [SR-IOV Network Device Plugin](https://github.com/k8snetworkplumbingwg/sriov-network-device-plugin) that discovers and advertises SR-IOV Virtual Functions (VFs) available on a Kubernetes node, and is used to map VFs to scheduled pods. To use it, a CNI meta plugin, supporting network provisioning based on device plugin (like [Multus CNI](https://github.com/k8snetworkplumbingwg/multus-cni) or [DANM](https://github.com/nokia/danm)) and SR-IOV CNI, to enable additional network interfaces for VFs (beyond the original network interface for Kubernetes control), are required. During pod creation, plumbs allocated SR-IOV VF to a pods network namespace using VF information given by the meta plugin. On pod deletion, CNI plugin releases the VF from the pod.
```
Comment (pgoyal01) [inline doc/ref_arch/kubernetes/chapters/chapter03.md:223]:
```suggestion
An example of often used device plugin is [SR-IOV Network Device Plugin](https://github.com/k8snetworkplumbingwg/sriov-network-device-plugin) that discovers and advertises SR-IOV Virtual Functions (VFs) available on a Kubernetes node, and is used to map VFs to scheduled pods. To use it, a CNI meta plugin (such as [Multus CNI](https://github.com/k8snetworkplumbingwg/multus-cni) or [DANM](https://github.com/nokia/danm)), to enable additional network interfaces for VFs (beyond the original network interface for Kubernetes control), and a SR-IOV CNI is required. During pod creation, plumbs allocated SR-IOV VF to a pods network namespace using VF information given by the meta plugin. On pod deletion, CNI plugin releases the VF from the pod.
```

**Comment by walterkozlowski** [comment]
@petorre and @pgoyal01 - are the issues with out of date diagrams resolved? If yes, @pgoyal01 please approve, of not - @petorre please update, and I will be happy to merge.

File: doc/ref_arch/kubernetes/chapters/chapter03.md [hunk 5]
```diff
@@ -211,6 +211,17 @@ Memory or Huge Pages are not considered by the Topology Manager. This can be don
 
 [Device Plugin Framework](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/) advertises device hardware resources to kubelet with which vendors can implement plugins for devices that may require vendor-specific activation and life cycle management, and securely maps these devices to containers.
 
+Figure 3-2 shows in four steps how device plugins operate on a Kubernetes node:
+* 1: During setup, the cluster administrator (more in [3.2.1.8 Operator Pattern](chapter03.md#3218-operator-pattern)) knows or discovers (as per [3.2.1.5 Node Feature Discovery](chapter03.md#3215-node-feature-discovery)) what kind of devices are present on the different nodes, selects which devices to enable and deploys the associated device plugins.
```
Comment (rgstori) [inline doc/ref_arch/kubernetes/chapters/chapter03.md:215]:
```suggestion
* 1: During setup, the cluster administrator (more in [3.2.5.1 Operator Pattern](chapter03.md#3251-operator-pattern)) knows or discovers (as per [3.2.1.5 Node Feature Discovery](chapter03.md#3215-node-feature-discovery)) what kind of devices are present on the different nodes, selects which devices to enable and deploys the associated device plugins.
```
fixed

File: doc/ref_arch/kubernetes/chapters/chapter03.md [hunk 6]
```diff
@@ -211,6 +211,17 @@ Memory or Huge Pages are not considered by the Topology Manager. This can be don
...
+* 4: Kubelet decides which node to schedule workload on and which device on the node to assign to the pod's containers.
+
+<p align="center"><img src="../figures/Ch3_Figure_Device_Plugin_operation.png" alt="Device Plugin Operation" Title="Device Plugin Operation" width="50%"/></p>
+<p align="center"><b>Figure 3-2:</b> Device Plugin Operation</p>
+
+An example of often used device plugin is [SR-IOV Network Device Plugin](https://github.com/k8snetworkplumbingwg/sriov-network-device-plugin) that discovers and advertises SR-IOV Virtual Functions (VFs) available on a Kubernetes node, and is used to map VFs to scheduled pods. To use it, a CNI meta plugin (such as [Multus CNI](https://github.com/k8snetworkplumbingwg/multus-cni) or [DANM](https://github.com/nokia/danm)), to provision additional network interfaces for VFs (beyond the original network interface for Kubernetes control), and a SR-IOV CNI is required. During pod creation, plumbs allocated SR-IOV VF to a pods network namespace using VF information given by the meta plugin. On pod deletion, CNI plugin releases the VF from the pod.
```
Comment (rgstori) [inline doc/ref_arch/kubernetes/chapters/chapter03.md:223]:
```suggestion
An example of often used device plugin is the [SR-IOV Network Device Plugin](https://github.com/k8snetworkplumbingwg/sriov-network-device-plugin), that discovers and advertises SR-IOV Virtual Functions (VFs) available on a Kubernetes node, and is used to map VFs to scheduled pods. To use it, the SR-IOV CNI is required, as well as a CNI multiplexer plugin (such as [Multus CNI](https://github.com/k8snetworkplumbingwg/multus-cni) or [DANM](https://github.com/nokia/danm)), to provision additional secondary network interfaces for VFs (beyond the primary network interface). During pod creation, the plugin allocates a SR-IOV VF to a pod's network namespace using the VF information given by the meta plugin. On pod deletion, the CNI plugin releases the VF from the pod.
```
clarification

**Review by rgstori — COMMENTED** [review]
approved pending suggested fixes

File: doc/ref_arch/kubernetes/chapters/chapter03.md [hunk 7]
```diff
@@ -212,15 +212,15 @@ Memory or Huge Pages are not considered by the Topology Manager. This can be don
...
+An example of often used device plugin is the [SR-IOV Network Device Plugin](https://github.com/k8snetworkplumbingwg/sriov-network-device-plugin), that discovers and advertises SR-IOV Virtual Functions (VFs) available on a Kubernetes node, and is used to map VFs to scheduled pods. To use it, the SR-IOV CNI is required, as well as a CNI multiplexer plugin (such as [Multus CNI](https://github.com/k8snetworkplumbingwg/multus-cni) or [DANM](https://github.com/nokia/danm)), to provision additional secondary network interfaces for VFs (beyond the primary network interface). During pod creation, the plugin allocates a SR-IOV VF to a pod's network namespace using the VF information given by the meta plugin. On pod deletion, the CNI plugin releases the VF from the pod.
```
Comment (pgoyal01) [inline doc/ref_arch/kubernetes/chapters/chapter03.md:223]: In the last sentence can we clarify which plugin as the paragraph references the SR-IOV and the multiplexer plugins.

**Review by CsatariGergely — APPROVED** [review]

**Review by tomkivlin — APPROVED** [review]

File: doc/ref_arch/kubernetes/chapters/chapter03.md (hunk 7 above)
Comment (petorre) [reply doc/ref_arch/kubernetes/chapters/chapter03.md]: > In the last sentence can we clarify which plugin as the paragraph references the SR-IOV and the multiplexer plugins.

@pgoyal01 : Please see if current text clarifies it.

**Comment by petorre** [comment]
> @petorre and @pgoyal01 - are the issues with out of date diagrams resolved? If yes, @pgoyal01 please approve, of not - @petorre please update, and I will be happy to merge.

@walterkozlowski : I'm not sure which out of date diagrams?! This PR adds one [figure](https://raw.githubusercontent.com/cntt-n/CNTT/d2bc2fc4780acd8596975332cd9fe0904e81ff64/doc/ref_arch/kubernetes/figures/Ch3_Figure_Device_Plugin_operation.png) on Plugin operation 4 steps.

File: doc/ref_arch/kubernetes/chapters/chapter03.md (hunk 7 above)
Comment (pgoyal01) [reply doc/ref_arch/kubernetes/chapters/chapter03.md]: @petorre Thanks.

**Review by pgoyal01 — APPROVED** [review]

File: doc/ref_arch/kubernetes/chapters/chapter03.md [hunk 8]
```diff
@@ -211,6 +211,17 @@ Memory or Huge Pages are not considered by the Topology Manager. This can be don
 
 [Device Plugin Framework](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/) advertises device hardware resources to kubelet with which vendors can implement plugins for devices that may require vendor-specific activation and life cycle management, and securely maps these devices to containers.
 
+Figure 3-2 shows in four steps how device plugins operate on a Kubernetes node:
+* 1: During setup, the cluster administrator (more in [3.2.5.1 Operator Pattern](chapter03.md#3251-operator-pattern)) knows or discovers (as per [3.2.1.5 Node Feature Discovery](chapter03.md#3215-node-feature-discovery)) what kind of devices are present on the different nodes, selects which devices to enable and deploys the associated device plugins.
+* 2: The plugin reports the devices it found on the node to the Kubelet device manager and starts its gRPC server to monitor the devices.
+* 3: A user submits a pod specification (workload manifest file) requesting a certain type of device.
+* 4: Kubelet decides which node to schedule workload on and which device on the node to assign to the pod's containers.
```
Comment (gkunz) [inline doc/ref_arch/kubernetes/chapters/chapter03.md:218]: nit-pick: the scheduler decides the node and kubelet handles the device allocation.
```suggestion
* 4: The scheduler determines a suitable node based on device availability and the local kubelet assigns a specific device to the pod's containers.
```

Comment (petorre) [reply doc/ref_arch/kubernetes/chapters/chapter03.md]: > nit-pick: the scheduler decides the node and kubelet handles the device allocation.

@gkunz : Thank you!

**Review by gkunz — APPROVED** [review]

**Review by rgstori — APPROVED** [review]

**Pull request merged**